File: 1044-find-common-characters/find-common-characters.py
class Solution:
    def count(self, word):
        frequency = [0] * 26
        for char in word:
            frequency[ord(char) - ord('a')] += 1
        return frequency

    def intersection(self, freq1, freq2):
        return [min(f1, f2) for f1, f2 in zip(freq1, freq2)]

    def commonChars(self, words: List[str]) -> List[str]:
        last = self.count(words[0])

        for word in words[1:]:
            last = self.intersection(last, self.count(word))
        
        result = []
        for i in range(26):
            result.extend([chr(i + ord('a'))] * last[i])
        
        return result

File: 1044-find-common-characters/find-common-characters.js
/**
 * @param {string[]} words
 * @return {string[]}
 */
function count(word) {
    const frequency = Array(26).fill(0);
    for (let char of word) {
        frequency[char.charCodeAt(0) - "a".charCodeAt(0)]++;
    }
    return frequency
}

function intersection (freq1, freq2) {
    return freq1.map((val, indx) => Math.min(val, freq2[indx]));
}


var commonChars = function(words) {
    let last = count(words[0])
    for (let i = 1; i < words.length; i++) {
        last = intersection(last, count(words[i]))
    }

    const result = [];
    for (let i = 0; i < 26; i++) {
        while (last[i] > 0) {
            result.push(String.fromCharCode(i + 'a'.charCodeAt(0)));
            last[i]--;
        }
    }

    return result;
};